src/core/midiPorts.cpp

```diff
@@ -0,0 +1,426 @@
+{
+	// TODO: Compatibility with legacy code - to be removed later on // 
+	if (msg->size() < 3) {
+		return;
+	}
+	midiDispatcher::dispatch(msg->at(0), msg->at(1), msg->at(2));
+	// TODO: End of compatibility with legacy code // 
+
+
+	// "data" pointer should point at a port name string // 
+	// std::string port = *static_cast<std::string*>(data);
+
+	// A port name shall be converted into the address string
+	// port = "p;" + port;
+
+	// Creating a nice MidiMsg instance //
+	// MidiMsg mm = MidiMsg(port, msg);
+
+	// Passing it to a new midiDispatcher - TODO //
+	// something like:
+	// midiDispatcher::dispatch(mm);
+}
+
+// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+
+// TODO: Legacy function - to be removed in next stages of overhaul
+void sendMidiLightningInitMsgs_()
+{
+	for (const midimap::Message& m : midimap::midimap.initCommands) {
+		if (m.value != 0x0 && m.channel != -1) {
+			u::log::print("[KM] MIDI send (init) - Channel %x - Event 0x%X\n", m.channel, m.value);
+			MidiEvent e(m.value);
+			e.setChannel(m.channel);
+			kernelMidi::send(e.getRaw());
+		}
+	}
+}
+
+} // {anonymous}
+
+
+/* -------------------------------------------------------------------------- */
+
+void init()
+{
+	setApi(conf::conf.midiSystem);
+	midiOut_ = new RtMidiOut((RtMidi::Api) api_, "Giada dummy port");
+	midiIn_  = new RtMidiIn((RtMidi::Api) api_, "Giada dummy port");
+
+	// TODO: Open In port and out port for Stage 1
+}
+
+// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+
+bool hasAPI(int API)
+{
+	std::vector<RtMidi::Api> APIs;
+	RtMidi::getCompiledApi(APIs);
+	for (unsigned i=0; i<APIs.size(); i++)
+		if (APIs.at(i) == API)
+			return true;
+	return false;
+}
+
+// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+
+void setApi(int api)
+{
+	api_ = api;
+	u::log::print("[MP] using system 0x%x\n", api_);
+}
+
+/* -------------------------------------------------------------------------- */
+
+std::vector<std::string> getOutDevices(){
+	
+	std::vector<std::string> output;
+
+	int count;
+	try {
+		// Get port count first //
+		count = midiOut_->getPortCount();
+	
+		// Get names of all available ports //
+		for (int i=0; i<count; i++) {
+			output.push_back(midiOut_->getPortName(i));
+		}
+	}
+	catch (RtMidiError &error) {
+		u::log::print("[MP::getOutDevices] RtMidiError: %s\n",
+						error.getMessage().c_str());
+	}
+
+	return output;
+}
+
+// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+
+std::vector<std::string> getInDevices(){
+	
+	std::vector<std::string> output;
+
+	int count;
+	try {
+		// Get port count first //
+		count = midiIn_->getPortCount();
+
+		// Get names of all available ports //
+		for (int i=0; i<count; i++) {
+			output.push_back(midiIn_->getPortName(i));
+		}	
+	}
+	catch (RtMidiError &error) {
+		u::log::print("[MP::getInDevices] RtMidiError: %s\n",
+						error.getMessage().c_str());
```

Comment:
Is there a way to improve the API of `u::log::print`? I have a `.c_str()` overdose... :-)

Reply:
I feel your pain. The `u::log::print` function is based on `printf` so there's nothing much you can do about it... Unless we switch to a more powerful logging system, or base the current one on [fmt](https://github.com/fmtlib/fmt) or similar.

Reply:
That'd be awesome!